main.go

```diff
@@ -0,0 +1,273 @@
+	}
+
+	var c int
+	var keys []string
+	var muts []*bigtable.Mutation
+
+	for rows.Next() {
+		var key, cf, col string
+		var value []byte
+		var ts int64
+		var chk uint32
+		if err := rows.Scan(&key, &cf, &col, &value, &ts, &chk); err != nil {
+			return fmt.Errorf("failed to scan table-list results: %s", err)
+		}
+
+		timestamp := time.Unix(0, ts)
+
+		hasher := fnv.New32a()
+		_, _ = fmt.Fprintf(hasher, key)
+		_, _ = fmt.Fprintf(hasher, cf)
+		_, _ = fmt.Fprintf(hasher, col)
+		_, _ = hasher.Write(value)
+		_, _ = fmt.Fprintf(hasher, timestamp.Format(time.RFC3339Nano))
+
+		if computed := hasher.Sum32(); chk != computed {
+			return fmt.Errorf("integrity check failed, db.chk=%d, computed hash = %d", chk, computed)
+		}
+
+		m := bigtable.NewMutation()
+		m.Set(cf, col, bigtable.Time(timestamp), value)
+
+		c++
+		keys = append(keys, key)
+		muts = append(muts, m)
+
+		// nothing magic about 100, tune as necessary - we just need some limit so that grpc payloads don't get too huge
+		if len(muts) == 100 {
+			if errs, err := tbl.ApplyBulk(ctx, keys, muts); err != nil {
+				return fmt.Errorf("failed to write to bigtable: %s", err)
+			} else if errs = filterErrors(errs); len(errs) > 0 {
+				return fmt.Errorf("failed to write to bigtable, %d errors, first: %s", len(errs), errs[0])
+			}
+
+			keys = nil
+			muts = nil
+		}
+	}
+
+	if len(muts) > 0 {
+		if errs, err := tbl.ApplyBulk(ctx, keys, muts); err != nil {
+			return fmt.Errorf("failed to write to bigtable: %s", err)
+		} else if errs = filterErrors(errs); len(errs) > 0 {
+			return fmt.Errorf("failed to write to bigtable, %d errors, first: %s", len(errs), errs[0])
+		}
+	}
+
+	log.Printf("restored %d rows for %s", c, table)
+	return nil
+}
+
+func saveAll(ctx context.Context, db *sql.DB, adminClient *bigtable.AdminClient, btClient *bigtable.Client) error {
+	tables, err := adminClient.Tables(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to list BT tables: %s", err)
+	}
+
+	sort.Strings(tables)
+
+	for _, table := range tables {
+		if err := saveTable(ctx, table, db, btClient); err != nil {
+			return fmt.Errorf("failed to save table %q: %s", table, err)
+		}
+	}
+
+	return nil
+}
+
+func saveTable(ctx context.Context, table string, db *sql.DB, btClient *bigtable.Client) error {
+	log.Printf("saving %q table...", table)
+	stmt, err := db.Prepare(`DROP TABLE IF EXISTS "` + table + `"`)
+	if err != nil {
+		return fmt.Errorf("failed to prepare DROP TABLE: %s", err)
+	}
+	if _, err := stmt.Exec(); err != nil {
+		return fmt.Errorf("failed to execute DROP TABLE: %s", err)
+	}
+
+	stmt, err = db.Prepare(`CREATE TABLE IF NOT EXISTS "` + table + `" (key TEXT, column_family TEXT, column TEXT, value BLOB, timestamp INTEGER, chk INTEGER)`)
+	if err != nil {
+		return fmt.Errorf("failed to prepare CREATE TABLE: %s", err)
+	}
+	if _, err := stmt.Exec(); err != nil {
+		return fmt.Errorf("failed to execute CREATE TABLE: %s", err)
+	}
```

Comment:
I see.. so there are no unique keys or ids or anything, which means that flushing out a memory snapshot entails completely nuking the last snapshot and doing a full dump.

How often (in local dev) will we invoke this tool to save?

Reply:
I dunno... every 5s maybe?  Probably could be quite a bit less often than that and it would still be fine...